feat(schema): application syntax draft/2-0-0

eeberhard opened the pull request:
<!-- Pull Request guidelines:
1. Give the PR a relevant and descriptive title, using one of the supported commit types:
   [ build, ci, chore, docs, feat, fix, perf, refactor, release, revert, style, test ]
2. Link the PR to the parent issue, which should already describe and motivate the PR
3. Explain how the PR addresses the parent issue
4. Add any supporting resources (screenshots, test results, etc)
5. Help the reviewer by suggesting the method and estimated time to review
6. If applicable, make a checklist of tasks that should be completed before merging
7. If applicable, mention related issues (blocked by / blocks)
8. Post creation actions: tag reviewers and link the PR to its parent issue under the Development section
-->

## Description

<!-- Required: link the parent issue -->

- #147 

<!-- Optional: add additional resources (links, screenshots, test results, etc)
## Supporting information
-->

## Review guidelines

<!-- Required: estimate how long a review should take -->
Estimated Time of Review: 90 minutes

Review by commit

## Checklist before merging:

- [x] Confirm that the relevant changelog(s) are up-to-date in case of any user-facing changes

<!-- Optional: define further tasks that should be completed before merging
- [x] Task 1
- [ ] Task 2
-->

<!-- Optional: link related issues
## Related issues
### Blocked by:
- #0

### Blocks:
- #0
-->

eeberhard commented:
Example of the new syntax:

```yaml
# yaml-language-server: $schema=https://raw.githubusercontent.com/aica-technology/api/feat/draft-v2.0.0/docs/static/schemas/draft/2-0-0/application.schema.json

schema: draft/2-0-0

dependencies:
  base: v4.0.0
  packages:
    - { package: "@aica/components/pose-detection", version: "v1.0.0" }
    - { package: "@aica/collections/ur-collection", version: "v3.1.0", constraints: ">= 3.0.0" }

metadata:
  name: My application
  description: Some description of the application
  tags:
    - "motion"
    - "force control"
    - "universal robots"

userdata:
  info: "..."
  website: github.com/eeberhard

on_start:
  load:
    - component: my_component
    - hardware: my_hardware
  sequence:
    start: my_sequence

conditions:
  my_condition:
    display_name: My cool condition
    condition:
      any:
        - component: foo
          state: active
        - controller: foo
          hardware: bar
          predicate: baz
    events:
      unload:
        component: foo

sequences:
  my_sequence:
    display_name: My cool sequence
    steps:
      - delay: 10
      - check:
          condition:
            condition: my_condition
          else:
            load:
              component: foo
      - check:
          condition:
            component: foo
            state: active
          wait_forever: true
      - check:
          condition:
            sequence: foo
            state: active
          timeout: 10
          else:
            unload:
              component: foo

hardware:
  my_hardware:
    urdf: foo
    rate: 12.5    
    rate_tolerance: 0.99
    strict: true
    events:
      transitions:
        on_load:
          load:
            controller: my_controller
            hardware: my_hardware
    controllers:
      my_controller:
        plugin: my_controller/Foo
        events:
          transitions:
            on_load:
              switch_controllers:
                hardware: my_hardware
                activate: my_controller
          predicates:
            is_stable:
              sequence:
                start: my_sequence

components:
  foo:
    component: my_package::Foo
    parameters:
      a: foo
      b: bar
    inputs:
      my_input: /foo/my_input
    outputs:
      my_output: /foo/my_output
    events:
      transitions:
        on_configure:
          lifecycle: activate
      predicates:
        is_in_range:
          unload:
            component: foo

graph:
  positions:
    on_start:
      x: 20
      y: 40
    conditions:
      my_condition:
        x: 0
        y: 0
    sequences:
      my_sequence:
        x: 0
        y: 0
    components:
      foo:
        x: 0
        y: 0
    hardware:
      my_hardware:
        x: 0
        y: 0
```

domire8 commented on docs/static/schemas/draft/2-0-0/application.schema.json:
```diff
@@ -0,0 +1,1564 @@
+                      "type": "string"
+                    },
+                    "parameters": {
+                      "$ref": "#/$defs/parameters"
```
https://github.com/aica-technology/dynamic-state-engine/issues/278

Do we "allow" the rate parameter for controllers under the `parameters` field like we do for components or do we need to add it as a separate property?

eeberhard replied:
good point. If `rate` is a separate field for hardware, and we want that to apply to controllers and components equally, then I think it makes sense to have that as a separate optional field in each case, rather than assuming it under parameters.

domire8 commented on docs/static/schemas/draft/2-0-0/application.schema.json:
```diff
@@ -0,0 +1,1564 @@
+                    "on_configure": {
+                      "$ref": "#/$defs/events"
+                    },
+                    "on_activate": {
```
Why do we go from `is_active` to `on_activate`?

domire8 commented on docs/static/schemas/draft/2-0-0/application.schema.json:
```diff
@@ -0,0 +1,1564 @@
+                },
+                "switch_controllers": {
+                  "$ref": "#/$defs/events/properties/switch_controllers"
+                }
```
Need the set parameter event here

domire8 commented on docs/static/schemas/draft/2-0-0/application.schema.json:
```diff
@@ -0,0 +1,1564 @@
+                      "description": "The runtime state of the controller",
+                      "type": "string",
+                      "enum": [
+                        "unloaded",
```
Just to be sure, are those really all the states that the DSE can know about controllers? no inactive or configured?

domire8 reviewed (commented):
I reviewed mainly the generated schema file to get the high level understanding and I'm mostly happy!

1. It's a pity that now under `events` we need to have an additional property `transitions` or `predicates` since that increases the number of things you need to write and remember. No real solution to that because we need to differentiate between predicates and states...
2. I'm not sure I like the `transitions: on_load` syntax/choice of words in the hardware (I think it works better for components). To me, the `load` is not a transition and I think we should choose words that emphasize that whatever events you put under this block will happen **after** the load happened. This is because the `load` event of a hardware is blocking and not asynchronous (like it's the case for components). Just looking at 
```
hardware:
  my_hardware:
    urdf: foo
    rate: 12.5    
    rate_tolerance: 0.99
    strict: true
    events:
      transitions:
        on_load:
          load:
            controller: my_controller
            hardware: my_hardware
    controllers:
      my_controller:
        plugin: my_controller/Foo
        events:
          transitions:
            on_load:
              switch_controllers:
                hardware: my_hardware
                activate: my_controller
```
I'm thinking that you could run into race conditions here since it states that on load, we activate the controller but how can you be sure that the controller is already loaded.. It's very subtle and a bit hard to describe, we can discuss in person if it's not clear.

eeberhard commented:
Thanks for the feedback! Actually a lot of your points come down to this new way of triggering events from state transitions.

> now under events we need to have an additional property transitions or predicates 

As you said, we need to distinguish these two types of event drivers. I considered a flatter structure like this but I didn't like it as much:

```yaml
my_component:
  component: my_component::Foo
  transition_events:  # or just `transitions:`
    on_load: ...
  predicate_events:  # or just `predicates:`
    is_in_range: ...
```

> on load, we activate the controller but how can you be sure that the controller is already loaded..

Predicates trigger events on any rising edge. But the intention with state transition events is that they trigger an event iff the specific transition is completed. In the implementation of transition events, I would only trigger `on_activate` events after the component / controller / etc has actually been activated. Because the underlying engine already manages and monitors the states of all components, we can actually avoid race conditions with this pattern better than the current implementation.

I can see the confusion because in the current implementation, `on_load` is a fake predicate that is actually triggered asynchronously with component loading and does not actually guarantee that the component has been fully constructed. The other potentially confusing point is that a transition like `on_activate` looks a lot like the transition callback function in the modulo component SDK (which is partly intentional), but in the SDK that function is called _during_ the transition, not after. On that note...

> Why do we go from is_active to on_activate?

I wanted to use a recognizable and consistent "transition callback" language similar to the component SDK. The previous predicate formulation `is_active` triggered the event every time the component state changed from not active to active. For that particular transition it is equivalent. But now consider the `is_inactive` predicate, which could be triggered both after configuring or after deactivating. So the name of state transitions events should reflect the triggering transition like `on_configure` and `on_deactivate`.

You might call it `after_load`, `after_configure`, ... but to me the `on_` feels right.

> Are those really all the states [...] about controllers? no inactive or configured?

Actually controllers have full lifecycle states but are automatically configured and go to `inactive` when loaded. The only transitions we really control are loading, activating, deactivating and unloading (and thereby the only primary states the controller stays in are unloaded, inactive or active).

Part of the reason of this being a draft is to just try the implementation and see what states actually make sense to control and expose for hardware and controllers.

domire8 commented:
> As you said, we need to distinguish these two types of event drivers. I considered a flatter structure like this but I didn't like it as much

Agreed.

> Because the underlying engine already manages and monitors the states of all components, we can actually avoid race conditions with this pattern better than the current implementation.

Absolutely, and this will constitute a huge improvement obviously. I know all that but that underlying mechanism might not fully transparent to the user.

> I wanted to use a recognizable and consistent "transition callback" language similar to the component SDK

I agree about similar language but part of me still think it's too similar with the `on_activate`. A user that writes a component with the SDK has to know what to put in the `on_activate_callback` and then we tell him that if you put `on_activate` into the YAML app, it will be triggered **after** on activate. I don't know...

`after_` just doesn't look nice as a prefix :smile: I think we start with this version and see how we feel about it later

domire8 reviewed (approved):
> Part of the reason of this being a draft is to just try the implementation and see what states actually make sense to control and expose for hardware and controllers.

:fire:

yrh012 commented:
Very cool stuff! 🔥

One thing that we should consider is enforcing the jsonschema to not allow a list of load/lifecycle events in the schema.

domire8 commented:
> One thing that we should consider is enforcing the jsonschema to not allow a list of load/lifecycle events in a given sequence step.

Can you elaborate with an example? Is that something that's already not correct in the current schema or did we miss this in this PR?

domire8 commented:
This PR also addresses (and closes) issues
- https://github.com/aica-technology/api/issues/140
- https://github.com/aica-technology/dynamic-state-engine/issues/278
- https://github.com/aica-technology/dynamic-state-engine/issues/242

yrh012 commented:
> > One thing that we should consider is enforcing the jsonschema to not allow a list of load/lifecycle events in a given sequence step.
> 
> Can you elaborate with an example? Is that something that's already not correct in the current schema or did we miss this in this PR?
Here is an example with a small sequence that loads a list of components. In the current implementation, this is allowed in the schema but fails at execution indicating invalid load payload.

  prepare_components:
    # load the components
    - load: 
        - component: assembled_frame_broadcaster
        - component: approach_frame_broadcaster
        - component: approach_tf_to_signal

eeberhard commented:
> This PR also addresses issues [...]

Thanks @domire8 !

> One thing that we should consider is enforcing the jsonschema to not allow a list of load/lifecycle events in a given sequence step

I think this is something to discuss further in the parent issue #147, but as @yrh012 points out, the current schema allows a sequence step to contain a "multi-event structure". However, in the current backend implementation (using base v3.4.0 for example), a sequence step can only support a single event.

This basically means there is a discrepancy between the current schema and implementation, and the observation of that is more of a general bug / issue report. However the relevant thing to discuss for the point of syntax 2.0 is whether to [1] change the schema to match the implementation (restrict a sequence step to a single event) or [2] do not change the schema and instead fix the implementation to support a multi-event structure on a sequence step. The option [1] is a breaking syntax change and so would need t be part of 2.0.

eeberhard commented:
As discussed, I will merge this now such that we can start to roll out the draft/2-0-0 implementations. We can then make additional changes based on the comments as discussed before we officially release it.
